popovers: Allow popover menus to scale with font-size.

sayamsamal opened the pull request:
As part of the efforts to allow the UI to scale with the font-size, this PR makes changes in the popover-related css to allow for font-scaling in the popovers.

The base work for this was done by @karlstolley in #30745.

Fixes: Part of #30476.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

| Before | After |
|--------|--------|
| ![add-channel-before](https://github.com/user-attachments/assets/9d1b5d55-d10f-49cb-b10b-04fbb33b3df9) | ![add-channel-after](https://github.com/user-attachments/assets/29946551-1cb4-4c1f-a797-7ba8957923fa) |
| ![group-info-before](https://github.com/user-attachments/assets/9ed71926-2a7f-46cd-a626-318fa4c6cc04) | ![group-info-after](https://github.com/user-attachments/assets/e3f927e3-012f-4279-b960-8ef4fb9586be) |
| ![help-menu-before](https://github.com/user-attachments/assets/64c7cbbd-6641-4e69-8b0a-808bbf97bd69) | ![help-menu-after](https://github.com/user-attachments/assets/5d8be03c-866d-4d4d-abc5-4f587d36da6d) |
| ![left-sidebar-before](https://github.com/user-attachments/assets/bd05500b-3324-45bc-87ab-aebfe4d72ef2) | ![left-sidebar-after](https://github.com/user-attachments/assets/0827e9f1-22fd-461b-a747-e1d7f5405736) |
| ![mobile-send-before](https://github.com/user-attachments/assets/723b2ba7-dc03-4a24-b3a6-5efb7ea0351a) | ![mobile-send-after](https://github.com/user-attachments/assets/caeb2d2f-acb6-49a8-881c-23a5f1d81b0a) | 
| ![msg-actions-before](https://github.com/user-attachments/assets/cbf3f2a3-04f0-468c-b24b-7e6f17cdd091) | ![msg-actions-after](https://github.com/user-attachments/assets/9df69214-c4e5-4652-bdd0-e3d544440d6b) | 
| ![personal-menu-before](https://github.com/user-attachments/assets/f76e789f-74cc-45e4-9a2e-5b4efa8bfca9) | ![personal-menu-after](https://github.com/user-attachments/assets/3042036e-1c8a-4419-8244-ebe4aa916d6d) | 
| ![send-opts-before](https://github.com/user-attachments/assets/5fccf24d-9581-415a-89dd-2c3c9045dc69) | ![send-opts-after](https://github.com/user-attachments/assets/9f0b67ad-348c-443a-ab72-b6b147b0056c) | 
| ![stream-actions-before](https://github.com/user-attachments/assets/ddb516b0-83b4-4fed-a7ad-11b98f610301) | ![stream-actions-after](https://github.com/user-attachments/assets/1a9e802e-c42a-49c8-a6b0-297554b703dc) | 
| ![topic-actions-before](https://github.com/user-attachments/assets/ad7fa7a9-5c4e-43f2-bb1a-69b99ea83eb9) | ![topic-actions-after](https://github.com/user-attachments/assets/3fdd4f43-a09b-4b98-b52f-abef3fbe126a) | 
| ![topic-vis-before](https://github.com/user-attachments/assets/47a445c9-b0d7-4a5d-b3cf-279af92fa20e) | ![topic-vis-after](https://github.com/user-attachments/assets/cec68198-0536-4f16-939d-3cd95ce4b02c) | 

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

sayamsamal commented:
@amanagr can you give this PR a review? Thanks!

Here, in the table,
- Before == At 16/140 before this PR.
- After == At 16/140 after this PR.

sayamsamal commented on web/styles/popovers.css:
```diff
@@ -569,13 +569,13 @@ ul.popover-group-menu-member-list {
         text-decoration: none;
-        /* 15px at 14px/1em */
-        font-size: 1.0714em;
+        /* 15px at 22px/1em */
```
Reason for the change: `.modal_title` is at 22px.

sayamsamal commented on web/styles/popovers.css:
```diff
@@ -569,13 +569,13 @@ ul.popover-group-menu-member-list {
-        /* 19px at 14px/1em */
-        font-size: 1.3571em;
+        /* 19px at 22px/1em */
+        font-size: 0.8638em;
```
Reason for the change: `.modal_title` is at 22px.

sayamsamal commented on web/styles/popovers.css:
```diff
@@ -1077,8 +1077,8 @@ ul.popover-group-menu-member-list {
-        /* 15px at 14px/1em */
-        font-size: 1.0714em;
+        /* 15px at 16px/1em */
+        font-size: 0.9375em;
```
Reason for the change: `tab-picker` is at 16px.

sayamsamal commented on web/styles/popovers.css:
```diff
@@ -1135,7 +1135,7 @@ ul.popover-group-menu-member-list {
     .full-name {
         /* 18px at 14px/1em */
-        font-size: 1.2857;
+        font-size: 1.2em;
```
Reason for change: `.tippy-content` is at 15px. There's an error in the comment, will update soon to `/* 18px at 15px/1em */`

sayamsamal commented on web/styles/popovers.css:
```diff
@@ -1145,6 +1145,8 @@ ul.popover-group-menu-member-list {
 
     .user-type {
+        /* 14px at 15px/1em */
+        font-size: 0.9333em;
```
Reason for change: `.tippy-content` is at 15px.

alya commented:
This one needs more space; there's no reason not to give it as much width as we need to show the options.
![Screenshot 2024-07-12 at 12 43 32@2x](https://github.com/user-attachments/assets/cd0c266d-9939-4e07-8165-afcf204da87b)

And I guess we sometimes let it get bigger, so I dunno what's up:

![Screenshot 2024-07-12 at 12 45 12@2x](https://github.com/user-attachments/assets/b89e2f8f-8e2b-4353-9180-1da776a1eb1e)

alya commented:
Everything else looks good to me!

sayamsamal commented:
> This one needs more space; we should just give it as much width as we need to show the options. ![Screenshot 2024-07-12 at 12 43 32@2x](https://private-user-images.githubusercontent.com/2090066/348379351-cd0c266d-9939-4e07-8165-afcf204da87b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA4OTM1NzksIm5iZiI6MTcyMDg5MzI3OSwicGF0aCI6Ii8yMDkwMDY2LzM0ODM3OTM1MS1jZDBjMjY2ZC05OTM5LTRlMDctODE2NS1hZmNmMjA0ZGE4N2IucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MTNUMTc1NDM5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OGVkOTZjYzhhZmJjZDQ5NzJmZDc0YTU4YjM0YjIwOTg4OWYyZGEzZmI2NDRhOTIwN2EwMWVkYWY1NWFkZGI5MSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.XmTPU5CVabUF42Q1oGw7PYw2RLO0TZcc0fcrAeKruHs)
> 
> And I guess we sometimes let it get bigger, so I dunno what's up:
> 
> ![Screenshot 2024-07-12 at 12 45 12@2x](https://private-user-images.githubusercontent.com/2090066/348379411-b89e2f8f-8e2b-4353-9180-1da776a1eb1e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA4OTM1NzksIm5iZiI6MTcyMDg5MzI3OSwicGF0aCI6Ii8yMDkwMDY2LzM0ODM3OTQxMS1iODllMmY4Zi04ZTJiLTQzNTMtOTE4MC0xZGE3NzZhMWViMWUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MTNUMTc1NDM5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Njc1ZmRhZmFlZmYwNjc4YWFhMjE1MjRlMDc0MWU3YmIxNjhkYzlhZGU3ZjYwZGEyY2E0MzI0OTk3OTJiMWQzMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.fGrJWo-WU7eop1c2QatIh1PjFARDTR2vHgl6s82bDbI)

This one doesn't seem to be related to this PR. I suspect this is because the space taken for the unread counter is not taken into account while calculating the width of the popover.

sayamsamal commented:
## Updates
- Fixed the issue when unread counts are present alongside the popover options.

| Before | After |
|--------|--------|
| ![Screenshot 2024-07-14 at 5 38 29 PM](https://github.com/user-attachments/assets/f453a7dd-49e3-4823-b5e3-01e34480400c) | ![Screenshot 2024-07-14 at 5 37 37 PM](https://github.com/user-attachments/assets/010864d7-1df6-4a46-95ae-83efdd662422) |
| ![Screenshot 2024-07-14 at 5 35 36 PM](https://github.com/user-attachments/assets/90e72d3e-b37e-4676-9350-e56ddaea3c13) | ![Screenshot 2024-07-14 at 5 34 16 PM](https://github.com/user-attachments/assets/0ed878ae-985f-4750-a751-4746fe86aec3) | 
- Added font-scaling for user card popover and user group popover.

| Before | After |
|--------|--------|
| ![Screenshot 2024-07-14 at 5 44 17 PM](https://github.com/user-attachments/assets/d4f958a0-ad11-41e8-8f31-5d5109d4f552) | ![Screenshot 2024-07-14 at 5 43 14 PM](https://github.com/user-attachments/assets/2f0ccef8-783b-45c8-bec5-983b68c80184) |
| ![Screenshot 2024-07-14 at 5 44 29 PM](https://github.com/user-attachments/assets/838a9f68-9c7f-41ce-868b-e52922646a4c) | ![Screenshot 2024-07-14 at 5 43 47 PM](https://github.com/user-attachments/assets/2174ae59-12ff-4906-b949-e0c390935528) |

sayamsamal commented:
@alya could you please take another look at this? thanks!

alya commented:
Looks great to me!

alya commented:
@karlstolley Do you have any feedback?

karlstolley commented:
Overall this looks good! I'm going to test manually in a minute, but looking at the code, I think there's information in the commit messages and GitHub comments, especially regarding the em context, that probably belongs alongside the em calculations in the CSS.

It should be safe to squash cf76b92bfa with fixes my earlier commit, too. Thanks for catching those, @sayamsamal!

karlstolley commented:
Okay. Put this through some paces in manual testing, and I think it's in good shape. 

I did note that the private-stream icon slips out of alignment, which seems to get exaggerated at 16/140:

| Legacy | 16/140 |
| --- | --- |
| ![legacy-stream-popover](https://github.com/user-attachments/assets/c249c773-7ce1-479c-8939-923680c73022) | ![16-140-stream-popover](https://github.com/user-attachments/assets/7834fb05-3cd5-4674-870b-4016fb5c01d1) |
 
I note a number of places in popovers where items are spaced vertically with adjustments of a couple of pixels of margin, on classes like `.popover-menu .link-item`. I'm not sure if that's worth fixing for 9.0, but would certainly make an excellent post-release pass for helping to ensure that spacing is a) declared by some other mechanism than margin, probably, and b) scales as expected with the font-size.

alya commented:
Mm, yeah, would be good to shift that lock down into place -- it doesn't look great.

timabbott commented:
Merged, thanks @sayamsamal and @karlstolley! That lock bug should probably be tracked separately; it's present on `main` but seems pretty noticeable.

sayamsamal commented:
Opened #30921.
